Add a new ProfileChange notification type (fixes #4108)

ThibG opened the pull request:
This Pull Request introduces a new notification type for when people you follow change their name or profile picture.

There are still a few things to sort out:
- This is the first notification for transient information, whereas all previous notifications are backed by long-lived activity records that are still meaningful after every notification is discarded. For this reason, I have limited it to one profile change by followed user, but there is no automatic cleanup when all notifications are discarded. This could be improved.
- The UI only shows the new name/profile picture. This is far from ideal, but should be easy to change by someone familiar with Mastodon's web UI (I need help with that!). Ideally, I would like old and new avatar/name couples side by side with an horizontal arrow from the old one to the new one.

sorin-davidoi commented on app/javascript/mastodon/reducers/settings.js:
```diff
@@ -21,20 +21,23 @@ const initialState = ImmutableMap({
   notifications: ImmutableMap({
     alerts: ImmutableMap({
       follow: true,
+      profile_change: true,
```
I think you also need to add the key [here](https://github.com/tootsuite/mastodon/blob/master/app/javascript/mastodon/reducers/push_notifications.js#L7).

ThibG replied:
Thanks, I'll do so when I manage to get “normal” notifications working like I want first.

sorin-davidoi replied:
Sure thing. On the same note, [this](https://github.com/tootsuite/mastodon/blob/master/app/models/web/push_subscription.rb#L27) may also need to be changed, since `data['alerts'][notification.type.to_s]` may not exist.

ThibG replied:
Wouldnt that return nil, thus a false-y value, which is the default anyway?

ThibG replied:
Assuming ActiveRecord handles `:json` fields as `Hash` objects, this should be fine: https://ruby-doc.org/core-2.2.0/Hash.html#method-i-5B-5D

Also didn't run into any error while trying it out.

ThibG commented:
I've been fighting with the front-end codebase, and if I understand correctly, what I'm currently missing is:
- Include previous avatar URLs and previous `display_name` in the serialization (notification_serializer.rb)
- Get the react.js controllers/containers/whatever to use them. Unfortunately, I really don't get that part, and I could use some help with this.

ThibG commented:
This almost works, but not quite: I can't figure out how to copy the avatar attachment from an `Account` to `ProfileChange`. Indeed, in this PR's current state, the *new* attachment ends up being saved.

We may be able to save the `ProfileChange` record before updating the `Account` one, but it is rather awkward, especially wrt. error/validation handling and factorization across the different ways of updating a profile (local profile, remote OStatus, remote ActivityPub).

nightpool commented on app/services/update_account_service.rb:
```diff
@@ -18,4 +23,8 @@ def authorize_all_follow_requests(account)
     end
   end
+
+  def notify_followers(account, previous_avatar, previous_display_name)
```
unused?

ThibG replied:
Indeed! Fixed!

ThibG commented:
Cleaned up a bit, added i18n strings for English and French, and tried in “production” (my single-user instance). Everything seems to work fine, although Desktop notifications and Web Push notifications are not nearly as useful since they only show the new account profile.

The code could certainly be refactored in a nicer way, but I'm not completely sure how.

The `ProfileChange` React.js containers and components could also be improved by people actually knowing CSS and React.js, but it's functional.

ThibG commented:
Confirmed to work with both OStatus and ActivityPub!

Outstanding issues:
- It's not really readable when “long” usernames are involved
- Duplicate `ProfileChange` records and notifications might be created (race conditions? Need to investigate further)
- `ProfileChange` records are created even if there are no followers to notify (also includes newly-discovered accounts)

nightpool commented:
So, just to be clear, this doesn't actually notify someone when the profile gets changed—only when their local server notices it?

ThibG commented:
@nightpool yes, but that's equivalent when dealing with ActivityPub (not with OStatus though), as Mastodon pushes profile changes to ActivityPub followers.

nightpool commented:
Ah, okay. but we'll only notify once per person? So if I change my display name twice it'll only notify for the first one? Or am I misunderstanding?

ThibG commented:
@nightpool only one notification will be saved per followed account, yeah. Followers will see only the last change.

nightpool commented:
I think I'm *tentatively* in favor of this change, as long as it's only
limited to display names and avatars.

It might make more sense to put it in the home feed though, as basically a
synthetic status? Not sure. I don't know how much I like putting things
that *don't directly relate to you* in your notifications
On Sat, Sep 23, 2017 at 12:10 AM Akihiko Odaki <notifications@github.com>
wrote:

> *@akihikodaki* requested changes on this pull request.
>
> Honestly I am not in favor of this change. Notifications can be noisy and
> the number should be minimized.
> I'd like to ask @Gargron <https://github.com/gargron> whether it is
> acceptable or not. I'll review the details if so.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tootsuite/mastodon/pull/4971#pullrequestreview-64731294>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAORV7kmIyhl-_pU9tbn3GIOjAqZTF7yks5slITKgaJpZM4PZzpw>
> .
>

ThibG commented:
@akihikodaki I guess it could be noisy, but profile changes are typically limited. I know I can't really generalize from my experience, but following ~70 people, I hardly even had a profile change that was not explicitly to try out that feature. It could also be disabled by default (not what my current feature request does).

@nightpool yes, it is limited to display names and avatar changes, I have no plan to extend it to any other kind of profile change. I can see how it could make sense to put it in the home feed, and I'm not opposed to that (not strongly in favor though). It would probably require a lot more work, however.

Psycojoker commented:
@akihikodaki @nightpool hello, you might want to read the original rational behind this idea https://github.com/tootsuite/mastodon/issues/4108

The short version is that for a non negligible group of users the avatar and/or the displayed name is the way to recognize other users in opposition to the username. Therefor, an unexpected change of avatar can results in a lot of confusion because the user can't be recognized anymore.

The group of users include people who have visual memories or dyslexic users has seen from the (enthusiastic) feedback gathered from sharing this idea, a real need has been expressed for that.

I think that adding a new kind of notification, that can be disabled or is disabled from the start, to make Mastodon a more accessible experience in general to its users is totally worth it and can totally make a difference :)

kyzh commented:
I concur with @psycojoker.

As a left to right reader, the profile picture is the main way to identify people.
Display name are a constant change for some accounts and actual handle are not often visible in full.
I'm personally always confused when people change their avatar.
Acknowledging this by having a setting, even hidden in the menu and off by default would be great way to improve my experience.

akihikodaki commented:
ThibG:
> I guess it could be noisy, but profile changes are typically limited.

kyzh:
> Display name are a constant change for some accounts and actual handle are not often visible in full.

I'm actually concerning accounts with such a constant change of their avatars or display names. Even such accounts are minor, it will make the notification ineffective for users with many followings and is not sustainable.

ThibG:
> It could also be disabled by default (not what my current feature request does).

If it is going to be merged, it should be enabled by default, or people just don't notice. This feature is helpful especially for inactive users, and enabling it by default is important to let such users take advantage of it.

ThibG commented:
Rebased on latest master.

@akihikodaki even if a user constantly changes their profile, there will be at most one notification for them (note, this is not completely true, as the client does not get notified of deleted notifications, similarly to what happens if someone repeatedly favorite/unfavorite a status, the notification will be sent out repeatedly even though it is stored only once).

I don't see many other options besides the Home Timeline thing, which has the advantage of not cluttering the notification column, but would require some very significant refactoring.

nightpool commented:
@Psycojoker i'm very well aware of the motivation. I just wanted to make sure I understood the implementation
